docs/Developers/NextGen_Core.rst

```diff
@@ -0,0 +1,303 @@
+
+        def generate_tasks(self):
+            result_length = sum(len(results)
+                                for results in self.dependency_results.values())
+            return (((i,), {}) for i in range(result_length))
+
+Running the Core
+----------------
+
+The `run(bears, result_callback, cache=None, executor=None)` method in
+``Core.py`` is used to set up a coala session and run it.
+- ``bears``: contains all the instances of bears to run.
+- ``result_callback``: a function that is called only when the results from
+the bears are available.
+- ``cache``: a cache used by bears to speed up runs.
+- ``executor``: a custom executor used to run the bears. By default
+``ProcessPoolExecutor`` is used.
+
+The new core comes with the following new additions:
+
+- **Dependency Resolution:** The ``initialize_dependencies`` method in
+  ``Core.py`` receives the bears that are to be run and processes bear
+  dependencies using a consumer-based system so that each dependency bear has
+  only one instance per section and file-dict. It returns a set of dependency
+  bears along with those bears that don't have any dependencies or whose
+  dependencies have been resolved (these are the ones that are scheduled to be
+  run). The bears are resolved by traversing each edge of the dependency graph
+  and instantiating dependency bears accordingly one per section. Before the
+  bears are run we initialize the dependency tracking in the ``__init__`` method
+  of the class ``Session`` which is responsible for running coala sessions. The
+  ``dependency_tracker`` object is used while scheduling and cleaning up bears
+  and also while finishing bear tasks.
+- **Task Scheduling:** The bears that have no dependencies or whose
+  dependencies have been resolved, only their tasks will be scheduled for
+  execution. Before executing any task coala looks it up in the cache.
+  In case of a hit, the existing results that are stored in the cache for the
+  corresponding task arguments are called using ``execute_task_with_cache``
+  method. In case of a miss or if coala is run without a cache the task is
+  executed. The bears without any running tasks are cleaned up from the state of
+  an ongoing run by resolving its dependencies, scheduling dependent bears
+  and removing the bear from the ``running_tasks`` dict. When a task belonging
+  to a particular bear completes, then ``_finish_task`` method is responsible
+  for checking whether the bear execution is complete and the result generated
+  from the task have been returned. If any new tasks are to be scheduled then
+  this method is responsible for doing so after the dependencies have been
+  resolved.
+- **result_callback :** Each of the bears scheduled to run for a session has
+  this callback function which is called when the results are available. It
+  should have the following signature ``def result_callback(result)``.
+- **executors:** These are instances of the ``ProcessPoolExecutor`` class
```

Comment:
Same for this one. Maybe we can move this to the part where you explain `executor` as a parameter of the `run()` in **Easier Interface** section. :-)

Reply:
Please read the original issue before reviewing. It clearly mentions that `result_callback` and `executors` belong in the `How to use core.run()` section. Since they are implementation details.

Reply:
https://github.com/coala/coala/issues/4584

Read the original comment :point_up:

Reply:
Yes the original issue comment mentions everything that you said in the last few comments. But I think that @Makman2 meant that comment to represent more of a checklist of things to cover in the tutorial than the **exact structure** of the tutorial. IMHO if we can make a few changes here and there to improve the readability we should do that.

But then again, its only my opinion. You and @Makman2 are the ones responsible to make decisions regarding this PR :)

Reply:
Actually much of the redundancy you described wasn't there in my first push but then as this PR grew, Makman asked for more sections to be included and more explanation for the same things in different places in the docs which now seems redundant to you but it has gone through several iteration. Still I appreciate your review and will keep it in mind if I have to change something before the final merge :+1:  Thanks.

Reply:
> I think that @Makman2 meant that comment to represent more of a checklist of things to cover in the tutorial than the exact structure of the tutorial. 

And that is correct, that wasn't intended to be the doc structure ;)

Reply:
And yeah, what's standing here should be incorporated into each feature-specific paragraph.